### PR TITLE
feat: move to poetry and remove setuptools

### PR DESCRIPTION
**Pull request by ahmet2mir**
This pull request is based on the issue #1021 

It:
* removes setuptools things (requirements.txt and setup.* files)
* create a make file to be able to run test locally (and not only on github actions)
* add [bandit](https://bandit.readthedocs.io/en/latest/) security checks
* initialize poetry environment, build, test, publish ...
* also added a [pex](https://pex.readthedocs.io/en/latest/) for "binary" file mode (which is more self extract archive)
* add editor config 
* restructure pyproject, extras "developer" and "debug" are for "dev" side and not user. I don't think that user will start to debug using pudb. And developper is not an extra but a separate section in pyproject. 

Also updated ci to use makefile and run tests on python 3.6 3.7 and 3.8 not only 3.8 plus split tests (branch) and release (tags) part

### CODE REVIEW

**Comment by vincentbernat** [comment]
Is poetry that widespread? For example, it is not available in Debian, which would make Exabgp not distributable in Debian/Ubuntu.

File: pyproject.toml [hunk 1]
```diff
@@ -42,28 +41,54 @@ classifiers = [
 
 [tool.poetry.dependencies]
 python = "~3.6 || ~3.7 || ^3.8"  # Compatible python versions must be declared here
+pyang = "^2.4.0"
+pyangbind = {git = "https://github.com/ahmet2mir/pyangbind.git"} # https://github.com/robshakir/pyangbind/pull/233/files
```
Comment (ahmet2mir) [inline pyproject.toml:45]: This is my fork to allow using python 3 where pyangbind uses enum which is not available for python3 https://github.com/ahmet2mir/pyangbind/blob/master/requirements.txt#L6

**Comment by ahmet2mir** [comment]
@vincentbernat yes poetry is not available as package on debian but poetry is only for "dev, test, build" part. Once packaged (wheel or tgz) poetry is not needed.

**Comment by thomas-mangin** [comment]
Hi @vincentbernat this is the one I wanted to tag ! Sorry for the noise !

**Comment by vincentbernat** [comment]
Poetry is not in Debian yet and since we package from the git repository, we need it. We can change the way we package and move away from the git repository (at the expense of some users who may build their packages from git too). However, I have doubt we would be able to build from the tgz either: will it contain a `setup.py`?

**Comment by ahmet2mir** [comment]
@vincentbernat setup.py is stil generated, for example after poetry build the targz contains a setup.py with

```
# -*- coding: utf-8 -*-
from setuptools import setup

package_dir = \
{'': 'src'}

packages = \
['exabgp',
 'exabgp.application',
 'exabgp.bgp',
 'exabgp.bgp.message',
 'exabgp.bgp.message.open',
 'exabgp.bgp.message.open.capability',
 'exabgp.bgp.message.update',
 'exabgp.bgp.message.update.attribute',
 'exabgp.bgp.message.update.attribute.bgpls',
 'exabgp.bgp.message.update.attribute.bgpls.link',
 'exabgp.bgp.message.update.attribute.bgpls.node',
 'exabgp.bgp.message.update.attribute.bgpls.prefix',
 'exabgp.bgp.message.update.attribute.community',
 'exabgp.bgp.message.update.attribute.community.extended',
 'exabgp.bgp.message.update.attribute.community.initial',
 'exabgp.bgp.message.update.attribute.community.large',
 'exabgp.bgp.message.update.attribute.sr',
 'exabgp.bgp.message.update.nlri',
 'exabgp.bgp.message.update.nlri.bgpls',
 'exabgp.bgp.message.update.nlri.bgpls.tlvs',
 'exabgp.bgp.message.update.nlri.evpn',
 'exabgp.bgp.message.update.nlri.qualifier',
 'exabgp.cli',
 'exabgp.conf',
 'exabgp.conf.yang',
 'exabgp.configuration',
 'exabgp.configuration.announce',
 'exabgp.configuration.core',
 'exabgp.configuration.flow',
 'exabgp.configuration.l2vpn',
 'exabgp.configuration.neighbor',
 'exabgp.configuration.operational',
 'exabgp.configuration.process',
 'exabgp.configuration.static',
 'exabgp.configuration.template',
 'exabgp.data',
 'exabgp.debug',
 'exabgp.environment',
 'exabgp.logger',
 'exabgp.netlink',
 'exabgp.netlink.route',
 'exabgp.protocol',
 'exabgp.protocol.ip',
 'exabgp.protocol.ip.tcp',
 'exabgp.protocol.iso',
 'exabgp.reactor',
 'exabgp.reactor.api',
 'exabgp.reactor.api.command',
 'exabgp.reactor.api.response',
 'exabgp.reactor.network',
 'exabgp.rib',
 'exabgp.util',
 'exabgp.vendoring']

package_data = \
{'': ['*']}

install_requires = \
['pyang>=2.4.0,<3.0.0',
 'pyangbind @ git+https://github.com/ahmet2mir/pyangbind.git@master',
 'pygments-yang-lexer>=0.2,<0.3']

entry_points = \
{'console_scripts': ['exabgp = exabgp.application.main:main',
                     'exabgp-cli = exabgp.application.cli:main',
                     'exabgp-healthcheck = exabgp.application.healthcheck:main',
                     'exabgpcli = exabgp.application.cli:main']}

setup_kwargs = {
    'name': 'exabgp',
    'version': '5.0.0',
    'description': 'BGP swiss army knife',
    'long_description': '...',
    'author': 'Thomas Mangin',
    'author_email': 'thomas.mangin@exa-networks.co.uk',
    'maintainer': None,
    'maintainer_email': None,
    'url': 'https://github.com/exa-networks/exabgp',
    'package_dir': package_dir,
    'packages': packages,
    'package_data': package_data,
    'install_requires': install_requires,
    'entry_points': entry_points,
    'python_requires': '>=3.6,<4.0',
}
setup(**setup_kwargs)
```
And personal opinion, I don't use system packages anymore to not rely on old unmaintained python packages from os vendor (hello redhat and scl, hello debian). I think that thoses packages should only be used for vendor things (like yum) and never interfer with them (even in el8 redhat emebed their own python for that purpose).
I prefer to setup as user [pipx](https://github.com/pipxproject/pipx) then install program via pipx like poetry, docker-compose,... then create [PEX](https://github.com/pantsbuild/pex) to use system python runtime only or [pyinstaller](https://github.com/pyinstaller/pyinstaller)/[pyoxidizer](https://github.com/indygreg/PyOxidizer) to embed the runtime. It makes the package distribution easier and faster (thanks to FPM). People could still use the package as dependency for their project but if they just wan't to use it no need to pip.

**Comment by vincentbernat** [comment]
So, we could move packaging outside the Git repository if we want to use Poetry.

As for system packages, this is an orthogonal issue. We don't maintain Debian packages just because we like it. Many users are asking for it because they don't want to deal with language-specific distribution systems or build custom packages. Old packages were new at some point and work good enough for them and they get uniformity and stability (as it "does not change", not as in "does not crash").

**Comment by thomas-mangin** [comment]
Hi @ahmet2mir thank you very much for your work, as I did not (and still does not) know poetry well and wanted to keep setuptool to be able to use my "trusted" release tooling, I had to manually incorporate your work into the repo, I also changed a few things but hopefully, this is now working on a local machine.

Merci et bonne continuation.

**Pull request closed without merging**